<launch>
  <arg name="side" />

  <group ns="$(arg side)_arm" clear_params="true">

    <!-- Load configuration -->
    <rosparam
      file="$(find koko_bringup)/config/$(arg side)_robot_parameters.yaml"
      command="load" />

    <!-- Broadcast our joint states as TFs -->
    <node
      name="robot_state_publisher"
      pkg="robot_state_publisher"
      type="state_publisher" />

    <!-- Load URDFs -->
    <param name="robot_description"
      command="$(find xacro)/xacro '$(find koko_descriptions)/robots/koko_$(arg side)_arm.urdf.xacro'" />

    <!-- Start communication with the motor drivers -->
    <node
      name="motor_interface"
      pkg="koko_hardware_drivers"
      type="bldc_driver.py"
      ns="koko_hardware"
      required="true"
      output="screen" />

    <!-- Start the control stack -->
    <include
      file="$(find koko_bringup)/config/controllers.xml"
      pass_all_args="true" />
    <node
      name="koko_controller_manager"
      pkg="koko_controller_manager"
      type="koko_controller_manager"
      output="screen" />
    <node name="koko_controller_spawner"
      pkg="koko_controller_manager"
      type="spawner.py"
      output="screen">
      <rosparam param="start_controllers">
        - koko_controllers/joint_state_controller
        - koko_controllers/gripper_controller
      </rosparam>
      <rosparam param="stop_controllers">
        - koko_controllers/cartesian_pose_controller
        - koko_controllers/joint_position_controller
      </rosparam>
    </node>

    <!-- Joint calibration -->
    <!-- TODO: this is temporary, and is being moved to the hardware interface initialization -->
    <node
      name="hardstop_calibrator"
      pkg="koko_bringup"
      type="hardstop_calibrate"
      output="screen" />

  </group>
</launch>
